### ISSUE TITLE
No ASM source bundles in SDK

### ISSUE THREAD

**Issue by iloveeclipse** [eclipse-platform/eclipse.platform.releng.aggregator#1767]
Just stumbled (in https://github.com/eclipse-pde/eclipse.pde/issues/1111) over the problem that ASM sources are not included into SDK, so developing PDE API tooling is harder as it should be (no Javadoc, no proper call hierarchy etc). Also same problem with JDT UI that uses ASM...

We should include src bundles for ASM libraries into SDK.

**Comment by iloveeclipse** [comment]
@laeubi : this is something that could be achieved with some clever tycho config, or that need to be a added as a dependency to some sources feature? Are there maven source bundles that can be used in Eclipse at all, or is it only possible with Orbit bundles?

**Comment by merks** [comment]
It would be unhelpful to point out that the target platform with the Oomph setup has the source:

![image](https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/assets/208716/e7b8b380-0b66-402e-bfbd-aa629d88f13f)

If you want sources in the actual SDK product, you should probably include the .source version of these features which I think will pull in the sources of all the requirements of those features.

![image](https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/assets/208716/190fddbb-4661-483f-a316-84c44d2964f4)

Note that sources have been gradually disappearing for 3rd party bundles that are no longer mentioned in (required by) any feature, and that has become the general trend to avoid locking in a specific version of such a dependency and to avoid redundancy of feature requirements that are already properly specified via requirements in bundles.

**Comment by laeubi** [comment]
It depends where asm is from and how the sdk is build, Tycho has an option to include all sources if they are available.

Beside that we added support to PDE to find sources if they are colocated so explicit inclusion should not be needed, also m2e can find sources if they are originating from a maven repository automatically.

**Comment by merks** [comment]
Is that include all sources for a product or a repo?  I assume the latter while @iloveeclipse wants the former.

**Comment by laeubi** [comment]
I never tried it for a product, but in case one can probably add support for products as well, or customize things on the product than on the feature level now we have mixed products.

**Comment by merks** [comment]
Well, I think products should include only what the requirements say they should include so that a director application can install the product from the repo, i.e., the publisher should produce the appropriate p2 metadata, but we digress...

**Comment by iloveeclipse** [comment]
Ed, do I understand you right, you proposed to add something like 
```
   <includes id="org.objectweb.asm.source" version="0.0.0"/>
```
to `eclipse.platform.releng.aggregator/eclipse.platform.releng/features/org.eclipse.sdk/feature.xml`?

And that would work without any pom magic?

**Comment by laeubi** [comment]
If it is really a product you can add `<plugin id="org.objectweb.asm.source" />` to the product as well ... adding it to the feature is not very optimal I think as it then will pull it in for everyone even if sources are not requested.

**Comment by iloveeclipse** [comment]
Well, in SDK I would expect everything needed to develop software. We ship our own platform sources with SDK too. Or do you mean, they are added via product?

Actually I think it is not only about ASM but whole bunch of 3rd party libraries we ship with SDK - I would naively expect to find their sources included too. Apache & co, bnd, whatever. If they are in SDK and we have code using it, so we should be able to debug that without doing anything extra, if we have SDK locally.

**Comment by laeubi** [comment]
As Ed pointed out the source are there "in the IDE", when building for example the udpate-site all sources are included:

https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/tree/master/eclipse.platform.releng.tychoeclipsebuilder/eclipse.platform.repository

so if I would want something specifically in a product (I assume you refere this one: https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/blob/master/eclipse.platform.releng.tychoeclipsebuilder/eclipse.platform.repository/sdk.product ) I would include it there, as you can see for some features the sources are already mentioned there.

This is not necessarily including any third party stuff as we stopped including them in the features directly.

**Comment by merks** [comment]
I gave very specific answer with a picture no less and a selection to show what to do where.  Here's it's in more detail with source feature dependencies added:

![image](https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/assets/208716/7f5dc37c-d4fc-4447-a6bc-8fa9ed54dfc2)

**Lots** of sources are missing as see in the installation profile, including source for these JDT UI things:

![image](https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/assets/208716/3f00708b-1858-4f5e-a11e-0b36766a8afe)

I believe the above would include the ASM source in the product, which is the title of this issue.  I made it clear several years ago, when we started down this path, that this would be the result, so I'm surprised to see this be noticed only now.  I'm not sure anyone should start listing a (huge) number of source bundle IDs in the product.  I know I don't want to maintain such a thing, and down the road, I expect we'd end up with things in the product that nothing actually needs...

**Comment by laeubi** [comment]
I think this is a more fundamental deployment problem rather than a configuration issue.

I therefore now created
- https://github.com/eclipse-equinox/p2/pull/446

to support the ground work for this and will add support in Tycho for this as well, then we can have an option for products as well similar to what we have for update sites to build products that include sources as much as possible.

**Comment by iloveeclipse** [comment]
> then we can have an option for products as well similar to what we have for update sites to build products

What it would mean for my use case: would it need only SDK product change to make SDK product "self contained" regarding sources without extra changes in any other features?

**Comment by merks** [comment]
> What it would mean for my use case: would it need only SDK product change to make SDK product "self contained" regarding sources without extra changes in any other features?

Yes, it would mean that with a single context property specified when using p2 director to create the installation, for any bundle that's installed, its corresponding source bundle, if available, would also be installed.  So one could remove all the includes of .source features everywhere and still end up with a product with corresponding source bundles for **everything**.  And, because these requirements are captured as p2 metadata, updates would work also exactly as one would expect, i.e., if any new bundle is updated or installed, it's corresponding source, if available, would also be installed.  That's the idea anyway....

**Comment by iloveeclipse** [comment]
> with a single context property specified when using p2 director to create the installation

I'm not p2 expert, so with that property added to some build file, would SDK contain all sources?

**Comment by laeubi** [comment]
There will be a new option in Tycho for the materialize products mojo similar to 
https://tycho.eclipseprojects.io/doc/latest/tycho-p2-director-plugin/materialize-products-mojo.html#installFeatures

**Comment by merks** [comment]
I expect Tycho would provide such a property via a convenience configuration boolean property, though technically it could be done via existing configuration support:

- https://tycho.eclipseprojects.io/doc/latest/tycho-p2-director-plugin/materialize-products-mojo.html#profileProperties
- https://tycho.eclipseprojects.io/doc/latest/tycho-p2-director-plugin/director-mojo.html#iuProfileproperties
